ateam_kenobi/src/skills/goalie.cpp

```diff
@@ -52,18 +59,207 @@ void Goalie::runFrame(
+  const Robot & goalie)
+{
+  const ateam_geometry::Segment goalie_line(
+    ateam_geometry::Point(-(world.field.field_length / 2) + 0.25, world.field.goal_width / 2),
+    ateam_geometry::Point(-(world.field.field_length / 2) + 0.25, -world.field.goal_width / 2));
 
   easy_move_to_.setTargetPosition(
     ateam_geometry::NearestPointOnSegment(
       goalie_line,
       world.ball.pos));
   easy_move_to_.face_point(world.ball.pos);
-  motion_commands.at(robot_id) = easy_move_to_.runFrame(maybe_robot.value(), world);
+  return easy_move_to_.runFrame(goalie, world);
+}
+
+ateam_msgs::msg::RobotMotionCommand Goalie::runBlockShot(const World & world, const Robot & goalie)
+{
+  const auto & ball_pos = world.ball.pos;
+  std::array<double, 16> distances;
+  std::transform(
+    world.their_robots.begin(), world.their_robots.end(), distances.begin(),
+    [&ball_pos](const auto & maybe_robot) {
+      if (!maybe_robot) {
+        return std::numeric_limits<double>::infinity();
+      }
+      return CGAL::approximate_sqrt(CGAL::squared_distance(ball_pos, maybe_robot->pos));
+    });
+  const auto min_distance_iter = std::min_element(distances.begin(), distances.end());
+  if (std::isinf(*min_distance_iter)) {
+    // Should only be true if there are no opponent robots
+    return runDefaultBehavior(world, goalie);
+  }
+  const auto opponent_id = std::distance(distances.begin(), min_distance_iter);
+  const auto opponent_pos = world.their_robots[opponent_id]->pos;
+  const auto opponent_ball_vector = world.ball.pos - opponent_pos;
+
+  const ateam_geometry::Ray shot_ray(opponent_pos, opponent_ball_vector);
+
+  const auto goalie_line_x = -(world.field.field_length / 2) + kRobotRadius;
+
+  const ateam_geometry::Segment goalie_line_extended(
+    ateam_geometry::Point(goalie_line_x, world.field.field_width / 2),
+    ateam_geometry::Point(goalie_line_x, -world.field.field_width / 2));
+
+  const ateam_geometry::Segment goalie_line(
+    ateam_geometry::Point(goalie_line_x, world.field.goal_width / 2),
+    ateam_geometry::Point(goalie_line_x, -world.field.goal_width / 2));
+
+  const auto maybe_intersection = CGAL::intersection(shot_ray, goalie_line_extended);
+
+  if (!maybe_intersection) {
+    return runDefaultBehavior(world, goalie);
+  }
+
+  ateam_geometry::Point shot_point_on_extended_goalie_line;
+
+  if (const auto * point = boost::get<ateam_geometry::Point>(&*maybe_intersection)) {
+    shot_point_on_extended_goalie_line = *point;
+  } else if (const auto * segment = boost::get<ateam_geometry::Segment>(&*maybe_intersection)) {
+    shot_point_on_extended_goalie_line = segment->source();
+  }
+
+  easy_move_to_.setTargetPosition(
+    ateam_geometry::NearestPointOnSegment(
+      goalie_line,
+      shot_point_on_extended_goalie_line));
+  easy_move_to_.face_point(world.ball.pos);
+
+  return easy_move_to_.runFrame(goalie, world);
+}
+
+ateam_msgs::msg::RobotMotionCommand Goalie::runBlockBall(const World & world, const Robot & goalie)
```

Comment:
We can probably refactor `runBlockBall` and `runBlockShot` a bit in the future since they share similar logic